Index: sstmacro-7.1.0/sstmac/main/sstmac.h
===================================================================
--- sstmacro-7.1.0.orig/sstmac/main/sstmac.h
+++ sstmacro-7.1.0/sstmac/main/sstmac.h
@@ -64,6 +64,9 @@ struct opts {
   bool print_params;
   bool low_res_timer;
   std::string cpu_affinity;
+  int shmid; // added by zhengyj
+  std::string go_shmid;
+  std::string go_file;
 
   opts() :
     help(0),
@@ -146,4 +149,4 @@ init_first_run(parallel_runtime* rt,
 
 }
 
-#endif
\ No newline at end of file
+#endif
